crates/ide_assists/src/handlers/add_turbo_fish.rs

```diff
@@ -77,26 +77,46 @@ pub(crate) fn add_turbo_fish(acc: &mut Assists, ctx: &AssistContext) -> Option<(
         }
     }
 
-    let number_of_arguments = generics.len();
-    let fish_head = std::iter::repeat("_").take(number_of_arguments).collect::<Vec<_>>().join(",");
+    let number_of_arguments = generics
+        .iter()
+        .filter(|param| {
+            matches!(param, hir::GenericParam::TypeParam(_) | hir::GenericParam::ConstParam(_))
+        })
+        .count();
 
     acc.add(
         AssistId("add_turbo_fish", AssistKind::RefactorRewrite),
         "Add `::<>`",
         ident.text_range(),
         |builder| match ctx.config.snippet_cap {
             Some(cap) => {
-                let snip = format!("::<${{0:{}}}>", fish_head);
+                let snip = format!("::<{}>", get_snippet_fish_head(number_of_arguments));
                 builder.insert_snippet(cap, ident.text_range().end(), snip)
             }
             None => {
+                let fish_head =
+                    std::iter::repeat("_").take(number_of_arguments).collect::<Vec<_>>().join(", ");
                 let snip = format!("::<{}>", fish_head);
                 builder.insert(ident.text_range().end(), snip);
             }
         },
     )
 }
 
+/// This will create a snippet string with tabstops marked
+fn get_snippet_fish_head(number_of_arguments: usize) -> String {
+    let mut fish_head = String::new();
+    let mut i = 1;
+    while i < number_of_arguments {
+        fish_head.push_str(&format!("${{{}:_}},", i));
+        i = i + 1;
+    }
```

Comment:
```suggestion
    let mut fish_head = String::new();
    for i in 1..number_of_arguments {
        fish_head.push_str(&format!("${{{}:_}},", i));
    }
```
which can be further simplified by using `Itertools`
```suggestion
    let mut fish_head = (1..number_of_arguments).format_with(", ", |i, f| {
        f(&format_args!("${{{}:_}},", i))
    });
```

Reply:
This is great, thanks! I've applied your suggestions